File: tests/r_parity/25_lmm.R
# StatsPAI LMM parity (R side) -- Module 25.
#
# Reads data/25_lmm.csv and runs lme4::lmer with REML.
# Tolerance: rel < 1e-6 on fixed effects, fixed-effect SEs, REML
# log-likelihood, and ICC.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(lme4)
})

MODULE <- "25_lmm"

df <- read_csv_strict(MODULE)
df$gid <- as.factor(df$gid)

fit <- lme4::lmer(y ~ x1 + (1 | gid), data = df, REML = TRUE)

co  <- fixef(fit)
ses <- sqrt(diag(vcov(fit)))
ll  <- as.numeric(logLik(fit))

# ICC from variance components.
vc <- as.data.frame(VarCorr(fit))
sigma2_group <- vc$vcov[vc$grp == "gid"]
sigma2_resid <- vc$vcov[vc$grp == "Residual"]
icc_val <- sigma2_group / (sigma2_group + sigma2_resid)

rows <- list(
  parity_row(MODULE, "beta_intercept",
             estimate = unname(co["(Intercept)"]),
             se = unname(ses["(Intercept)"]),
             n = nrow(df)),
  parity_row(MODULE, "beta_x1",
             estimate = unname(co["x1"]),
             se = unname(ses["x1"]),
             n = nrow(df)),
  parity_row(MODULE, "logLik", estimate = ll, n = nrow(df)),
  parity_row(MODULE, "icc", estimate = icc_val, n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(method = "REML",
                           n_groups = length(unique(df$gid))))

File: tests/r_parity/25_lmm.py
"""StatsPAI LMM parity (Python side) -- Module 25.

Random-intercept LMM on a deterministic n=1000, 50-group panel.
The companion 25_lmm.R uses lme4::lmer with REML. Tolerance:
rel < 1e-6 on fixed effects, fixed-effect SEs, REML log-likelihood,
and ICC after the StatsPAI REML reporting convention and optimizer
budget are aligned to lme4/Stata.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "25_lmm"


def make_panel(n_groups: int = 50, n_per: int = 20, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    gid = np.repeat(np.arange(n_groups), n_per)
    x1 = rng.normal(size=n_groups * n_per)
    group_re = rng.normal(scale=1.0, size=n_groups)
    y = (
        2.0 + 1.5 * x1 + group_re[gid]
        + rng.normal(scale=0.5, size=n_groups * n_per)
    )
    return pd.DataFrame({"y": y, "x1": x1, "gid": gid})


def main() -> None:
    df = make_panel()
    dump_csv(df, MODULE)

    fit = sp.mixed(data=df, y="y", x_fixed=["x1"], group="gid",
                   method="reml")

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "beta_intercept",
                     estimate=float(fit.params["_cons"]),
                     se=float(fit.std_errors["_cons"]),
                     n=int(fit.n_obs)),
        ParityRecord(MODULE, "py", "beta_x1",
                     estimate=float(fit.params["x1"]),
                     se=float(fit.std_errors["x1"]),
                     n=int(fit.n_obs)),
        ParityRecord(MODULE, "py", "logLik",
                     estimate=float(fit.log_likelihood),
                     n=int(fit.n_obs)),
        ParityRecord(MODULE, "py", "icc",
                     estimate=float(fit.icc),
                     n=int(fit.n_obs)),
    ]

    write_results(MODULE, "py", rows,
                  extra={"method": "reml",
                         "n_groups": int(fit.n_groups),
                         "cov_type": "unstructured"})


if __name__ == "__main__":
    main()
